spec/middleware-1.0.0-draft.1.md

```diff
@@ -0,0 +1,309 @@
+application frameworks provide such abstractions in the form of reusable
+software components.
+
+This specification provides a standardised signature for such components in the
+OWIN ecosystem, hereby referred to as **Middleware**.
+
+OWIN Middleware is defined in terms of a delegate structure building upon the
+OWIN: Open Web Server Interface for .NET [Owin 1.0], with no
+dependency on external libraries.
+
+### 1.1. Requirements Notation
+
+ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119][rfc2119].
+
+A component is deemed to be compliant with this specification if all absolute
+requirements and prohibitions are conformed to.
+
+Components and terms defined in this and other specifications are always
+capitalized, so as to clarify between a general term and the term as used in a
+specification.
+
+## 2. Definitions
+
+This document refers to the following terms:
+
+ - **Application**: a process or operation inspecting, modifying or otherwise
+   accessing HTTP messages, as an AppFunc signature.
+
+ - **Middleware**: Pass through components, capable of mediating request and
+   response messages, in a pipeline model which, when
+   composed, forms an Application.
+
+ - **Builder**: The extension point lambda used by developers to add Middleware
+   components to a pipeline.
+
+ - **Server**: As per the OWIN specification, an HTTP server that directly
+   communicates with the client, and uses OWIN semantics to process requests.
+
+## 3. Middleware
+
+An OWIN Application (as defined by an AppFunc) can be modelled as
+Middleware components layered in a pipeline.
+
+Each Middleware receiving a request may perform some processing on the request,
+may forward the request to the next Middleware and return the resulting
+response, or may decide to return a response immediately and not call the rest
+of the chain.
+
+In other words, a Middleware functions as a proxy of sort, completely
+encapsulating and hiding the next Middleware in the chain, transparently to the
+caller.
+
+### 3.1. Examples
+
+For example, a Middleware responsible for logging all requests and responses
+could execute as the first component of a pipeline, and log all requests and
+responses as they pass through.
+
+```
+        request      > (log request)      > (process request)
+SERVER =============== LOGGER ============= WEB APPLICATION
+        response     < (log response)     < (return response)
+```
+
+Other types of Middleware would respond immediately, without calling the next
+component in the pipeline, such as for an authorization module denying a request
+before a web application can respond.
+
+```
+        request      > (authenticate)
+SERVER =============== LOGGER ------X------ WEB APPLICATION
+        response     < (deny)
+```
+
+### 3.2. Middleware Operation
```

Comment:
Careful, we can't predict every possible usage of middleware to define what they can or cannot do. We can however list some best practices (e.g. SHOULD or MAY rather than MUST).

Some guidelines:
- A middleware SHOULD either produce a response OR call the next middleware.
- Middleware MAY alter the request fields and headers as needed before calling the next middleware, but SHOULD also revert these alterations afterwards as to avoid confusing upstream middleware.
- Middleware SHOULD be cautious of altering the response headers or writing to the body after calling the next middleware, as they may have already been sent to the client.
- The owin.ResponseHeaders and owin.ResponseBody entries MAY be replaced, but the component that did the replacement MUST eventually write to the original Dictionary and/or Stream. The Server is NOT expected to look for or utilize replacements in these fields, only the originals which it MUST maintain an internal reference to.